core/payment/src/api/provider.rs

```diff
@@ -326,8 +328,50 @@ async fn cancel_invoice(
     db: Data<DbExecutor>,
     path: Path<InvoiceId>,
     query: Query<Timeout>,
+    id: Identity,
 ) -> HttpResponse {
-    response::not_implemented() // TODO
+    let invoice_id = path.invoice_id.clone();
+    let node_id = id.identity;
+    let dao: InvoiceDao = db.as_dao();
+    let invoice = match dao.get(invoice_id.clone(), node_id).await {
+        Ok(Some(invoice)) => invoice,
+        Ok(None) => return response::not_found(),
+        Err(e) => return response::server_error(&e),
+    };
+
+    match invoice.status {
+        DocumentStatus::Issued => (),
+        DocumentStatus::Received => (),
+        DocumentStatus::Rejected => (),
+        DocumentStatus::Cancelled => return response::ok(Null),
+        DocumentStatus::Accepted | DocumentStatus::Settled | DocumentStatus::Failed => {
+            return response::conflict(&"Invoice already accepted by requestor")
+        }
+    }
+
+    with_timeout(query.timeout, async move {
+        match async move {
+            ya_net::from(node_id)
+                .to(invoice.recipient_id)
+                .service(BUS_ID)
+                .call(CancelInvoice {
+                    invoice_id: invoice_id.clone(),
+                    recipient_id: invoice.recipient_id,
+                })
+                .await??;
+            dao.cancel(invoice_id, node_id).await?;
+            Ok(())
+        }
+        .await
+        {
+            Ok(_) => response::ok(Null),
+            Err(Error::Rpc(RpcMessageError::Cancel(CancelError::Conflict))) => {
+                response::conflict(&"Invoice already accepted by requestor")
```

Comment:
this magic string is repeated 2x, maybe we can make it a const or a response type?

Reply:
it's not _magic_ in that sense that nowhere is it assumed that these two messages are identical and changing it doesn't break the code in any way. But I can extract it to a variable.

Reply:
True dat :)

Think you can leave it as is since its only 2x and in the same file...